lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -287,13 +287,20 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
                     $dqlAlias = $cacheKeyInfo['dqlAlias'];
                     $type     = $cacheKeyInfo['type'];
 
+                    if(
+                        isset($cacheKeyInfo['discriminatorColumn']) && 
+                        isset($data[$cacheKeyInfo['discriminatorColumn']]) &&
+                        $data[$cacheKeyInfo['discriminatorColumn']] != $cacheKeyInfo['discriminatorValue']
```

Comment:
This should be `!==` to avoid type juggling hell. Yes, that means that object discriminators will fail, that's an edge case that the ORM doesn't handle anyway.